[Functional Collective] enable custom work registration from python

yifuwang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #130583
* __->__ #130354

This PR does two things:
- Allow tensor -> work registration in Python via `torch._C._distributed_c10d.register_work`. Calling `torch.ops._c10d_functional.wait_tensor` on a tensor would trigger `.wait()` on the registered work object.
- Allow user-defined work object in Python to work with functional collectives.

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @chauhang

Commit:
[Functional Collective] enable custom work registration from python

[ghstack-poisoned]

Commit:
Update on "[Functional Collective] enable custom work registration from python"




cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu fegin XilunWu wanchaol fduwjj wz337 wconstab chauhang d4l3k

[ghstack-poisoned]

Commit:
Update on "[Functional Collective] enable custom work registration from python"


This PR does two things:
- Allow tensor -> work registration in Python via `torch._C._distributed_c10d.register_work`. Calling `torch.ops._c10d_functional.wait_tensor` on a tensor would trigger `.wait()` on the registered work object.
- Allow user-defined work object in Python to work with functional collectives.

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu fegin XilunWu wanchaol fduwjj wz337 wconstab chauhang d4l3k

[ghstack-poisoned]

Commit:
Update on "[Functional Collective] enable custom work registration from python"


This PR does two things:
- Allow tensor -> work registration in Python via `torch._C._distributed_c10d.register_work`. Calling `torch.ops._c10d_functional.wait_tensor` on a tensor would trigger `.wait()` on the registered work object.
- Allow user-defined work object in Python to work with functional collectives.

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu fegin XilunWu wanchaol fduwjj wz337 wconstab chauhang d4l3k

[ghstack-poisoned]

Commit:
Update on "[Functional Collective] enable custom work registration from python"


This PR does two things:
- Allow tensor -> work registration in Python via `torch._C._distributed_c10d.register_work`. Calling `torch.ops._c10d_functional.wait_tensor` on a tensor would trigger `.wait()` on the registered work object.
- Allow user-defined work object in Python to work with functional collectives.

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu fegin XilunWu wanchaol fduwjj wz337 wconstab chauhang d4l3k

[ghstack-poisoned]

wanchaol reviewed (approved):
lgtm, thanks for adding this!

wconstab commented on torch/csrc/distributed/c10d/PyProcessGroup.hpp:
```diff
@@ -41,6 +41,19 @@ class PyProcessGroup : public ProcessGroup {
+    }
+
+   private:
+    py::object py_obj_;
```
Work object lifetime is asynchronous with the life of the watchdog thread in C++.

Iirc we may not guarantee this thread exits before python shuts down. 

But if users call dist.destroy_process_group() as they are required to do, then the work objects should be finalized before python shuts down. 

Though, I don't think this PR does anything unsafe

Commit:
Update on "[Functional Collective] enable custom work registration from python"


This PR does two things:
- Allow tensor -> work registration in Python via `torch._C._distributed_c10d.register_work`. Calling `torch.ops._c10d_functional.wait_tensor` on a tensor would trigger `.wait()` on the registered work object.
- Allow user-defined work object in Python to work with functional collectives.

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu fegin XilunWu wanchaol fduwjj wz337 wconstab chauhang d4l3k

[ghstack-poisoned]

Commit:
Update on "[Functional Collective] enable custom work registration from python"


This PR does two things:
- Allow tensor -> work registration in Python via `torch._C._distributed_c10d.register_work`. Calling `torch.ops._c10d_functional.wait_tensor` on a tensor would trigger `.wait()` on the registered work object.
- Allow user-defined work object in Python to work with functional collectives.

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu fegin XilunWu wanchaol fduwjj wz337 wconstab chauhang d4l3k

[ghstack-poisoned]

yifuwang commented:
@pytorchbot merge
